sdk/python/feast/infra/offline_stores/contrib/postgres_offline_store/postgres.py

```diff
@@ -274,8 +274,10 @@ def to_sql(self) -> str:
     def _to_arrow_internal(self, timeout: Optional[int] = None) -> pa.Table:
         with self._query_generator() as query:
             with _get_conn(self.config.offline_store) as conn, conn.cursor() as cur:
-                conn.set_session(readonly=True)
+                conn.read_only = True
                 cur.execute(query)
+                if not cur.description:
+                    raise ZeroColumnQueryResult(query)
```

Comment:
I'm not necessarily against these error messages, but we probably want to do the same for all offline stores in that case. I'm also not sure what would actually trigger these errors.

Reply:
> I'm also not sure what would actually trigger these errors.

IIUC, these errors can be triggered when
1. The query returns zero rows.
2. The queried table doesn't have any columns in the table definition.

These errors will occur when using either one of `psycopg2` or `psycopg3`. I just did a small test. I spinned up a postgres database and created an empty table with no columns, and queries them using both versions.

```python
# psycopg2
>>> with conn.cursor() as cur: cur.execute("select * from empty_table;"); conn.commit(); cur.description
...
()
>>> with conn.cursor() as cur: cur.execute("select * from empty_table;"); res=cur.fetchone(); conn.commit(); print(res)
...
None

# psycopg3
>>> with conn.cursor() as cur: cur.execute("select * from empty_table;"); conn.commit(); cur.description
...
<psycopg.Cursor [TUPLES_OK] [INTRANS] (host=localhost database=postgres) at 0x15b6a8b40>
[]
>>> with conn.cursor() as cur: cur.execute("select * from empty_table;"); res=cur.fetchone(); conn.commit(); print(res)
...
<psycopg.Cursor [TUPLES_OK] [INTRANS] (host=localhost database=postgres) at 0x15b6a8880>
None

```

The only difference between the two implementations is that the type hints for `psycopg3` 'force' us to handle the case when `cur.description` or `cur.fetch()` (or a similar fetch operation) can be `None` or an empty list/tuple. For the `psycopg2` implementation the same situation can occur, however, the type hinting seems to be less strict, and thus the linter was not complaining. 

For full context, these are the errors that the linter is returning:
```
feast/infra/offline_stores/contrib/postgres_offline_store/postgres_source.py:118: error: Item "None" of "Optional[list[Column]]" has no attribute "__iter__" (not iterable)  [union-attr]
feast/infra/offline_stores/contrib/postgres_offline_store/postgres.py:281: error: Item "None" of "Optional[list[Column]]" has no attribute "__iter__" (not iterable)  [union-attr]
feast/infra/offline_stores/contrib/postgres_offline_store/postgres.py:344: error: Value of type "Optional[Any]" is not indexable  [index]
```

> I'm not necessarily against these error messages, but we probably want to do the same for all offline stores in that case

We could do the same for all offline stores. However, I think this is a peculiarity of psycopg2/3. Furthermore, I don't think these errors will actually be triggered ever, given the fact that the psycopg2 implementation seems to have almost identical behavior as psycopg3. 

Let me know what you think!

Reply:
Fair enough, ideally we'd have some sort of universal exception types that would span other stores as well, but these just seem like ways to suppress linter issues iiuc. It's unlikely that any realistic feast use path would actually hit those.

Reply:
Yes exactly! My thoughts as well :)